EOS-21433 cfgen: add support for FDMI filters

max-seagate opened the pull request:
Solution: add the code to configure the filters in CDF.

To use FDMI the following needs to be configured in CDF:

- FDMI plugin endpoint which will consume FDMI records sent by FDMI
  sources. Currently FDMI filter is one of Motr client processes;
- FDMI filters. `fdmi_filters` section defines filters and their FDMI
  plugin endpoints.

`cfgen/cfgen --help-schema` explains format of the `fdmi_filters`
section in CDF:

```
 # FDMI filter defines if an FDMI record would be sent from FDMI source to FDMI
 # plugin. Only M0_FDMI_FILTER_TYPE_KV_SUBSTRING filters are supported.
 #
fdmi_filters:  # This section is optional.  If it is missing, no fdmi filters
               # would be defined
  - name: <str>  # Human-readable name of the filter
    node: <str>  # Hostname of the node that hosts FDMI plugin
                 # that would receive every FDMI record this filter matches
    client_index: <int>  # 0-based index of the client that would be
                         # the FDMI plugin. Only "other" clients could be FDMI
                         # plugins. "s3" clients are not used to for this
                         # index.
                         # The client has to be defined in m0_clients section
                         # already.
    substrings: [ <str> ]  # List of strings for FDMI plugin to match
                           # See m0_conf_fdmi_filter::ff_substrings
                           # for the reference. May be empty or absent.
```

cfgen/examples/singlenode.yaml has an example of `fdmi_filters` section
commented. Uncomment it and run an FDMI plugin to start using an FDMI
filter.

If there is no fdmi_filters section then no fdmi filters would be added
to Motr configuration.

Signed-off-by: Maksym Medvied <maksym.medvied@seagate.com>

max-seagate commented:
Works only with https://github.com/Seagate/cortx-motr/pull/736. It's not based on `master` or `stable` because both are broken in my environment: https://github.com/Seagate/cortx-hare/issues/1628#issuecomment-854791080.

knekrasov commented on cfgen/cfgen:
```diff
@@ -1513,6 +1563,93 @@ class ConfProfile(ToDhall):
+        args = ', '.join([f'id = {oid_to_dhall(oid)}',
+                          f'rec_type = {self.rec_type}',
+                          f'filters = {oids_to_dhall(self.fdmi_filters)}'])
+        return '{ %s }' % args
```
```suggestion
        return f'{ {args} }'
```

max-seagate replied:
I did exactly as it was done in other the places, to keep the style uniform. I will make the suggested change though. Please let me know if it would be better to still keep the style uniform.

knekrasov commented on cfgen/cfgen:
```diff
@@ -1513,6 +1563,93 @@ class ConfProfile(ToDhall):
+        fdmi_flt_grp_id = new_oid(ObjT.fdmi_flt_grp)
+        fdmi_filter_oid = ConfFdmiFilter.build(
+            m0conf, parent, fdmi_flt_grp_id, fdmi_filter_desc, other_clients)
+        # XXX rec_type is M0_FDMI_REC_TYPE_FOL
```
I don't quite understand this comment. Is it something that we need to check in the code somehow (i.e. missing assert)?
Or it just explains magical `rec_type=0x1000`?

max-seagate replied:
Added better explanation.

knekrasov commented on cfgen/cfgen:
```diff
@@ -1513,6 +1563,93 @@ class ConfProfile(ToDhall):
+                          f'dix_pver = {oid_to_dhall(self.dix_pver)}',
+                          f'substrings = {substrings} : List Text',
+                          f'endpoints = ["{self.endpoint}"]'])
+        return '{ %s }' % args
```
```suggestion
        return f'{ {args} }'
```

max-seagate replied:
Done.

knekrasov commented on cfgen/cfgen:
```diff
@@ -406,6 +424,33 @@ def validate_profiles_desc(profiles_desc: List[Dict[str, Any]],
+            "doesn't refer to any node"
+        matching_nodes = [n for n in nodes if n.name == node]
+        assert len(matching_nodes) == 1, \
+            ("FDMI filter have exactly one node: {fdmi_filter_desc}",
```
```suggestion
            ("FDMI filter has exactly one node: {fdmi_filter_desc}",
```

knekrasov replied:
BTW, is this message rendered correctly? I can see that the string is not formatted (`f"someting"`) so `{fdmi_filter_desc}` will not be substituted.

max-seagate replied:
It's not. You're right.

Fixed here and in other places in the patch.

knekrasov commented on cfgen/cfgen:
```diff
@@ -112,6 +112,22 @@ profiles:  # This section is optional.  If it is missing, a single "default"
            # profile referring to all pools will be created.
   - name: <str>
     pools: [ <str> ]
+
```
Note: This fragment could also go to the PR description.

max-seagate replied:
I've mentioned how to get for the format format in the commit message. Will copy-paste and add it anyway.

knekrasov reviewed (changes requested):
There are a few style issues (probably minor ones) and the following general issues:
1. PR description is empty. Can you please add a description of the feature you've suggested? Links to other explanatory documents are highly appreciated.
2. PR description doesn't have a link to JIRA issue.
3. Commit message should be composed in the form as follows: "<JIRA-ID>: <problem you're solving with this commit>"
4. Commit description should describe the problem (if details are necessary) and describe the solution of the problem.

max-seagate commented:
Updated & rebased on `main`.

> PR description is empty. Can you please add a description of the feature you've suggested? Links to other explanatory documents are highly appreciated.

Poor GitHub UI. The description is there, in the commit message. Will have to copy-paste then.

> PR description doesn't have a link to JIRA issue.

Fixed.

> Commit message should be composed in the form as follows: ": <problem you're solving with this commit>"

Fixed. As I see in the `git log` this is not consistent across commits to `main`.

> Commit description should describe the problem (if details are necessary) and describe the solution of the problem.

Done.

max-seagate commented:
Rebased on the latest `main` and also fixed the following issue:

@knekrasov, MS teams:
```
... here are the tests that can help you to make sure that CDF generation (and hence Mini Provisioner) will work fine with your changes.
	
In Hare source folder, run make install or make devinstall
	Activate virtualenv like this: source ./.py3venv/bin/activate
	Go to the following sub-directory: cd provisioning/miniprov
	Run the tests python ./setup.py test (they will fail if Swapnil's concern is valid)
```

Now it's not failing in my code and it's failing somewhere else:
```
> python setup.py test 
running test
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
running egg_info
writing hare_mp.egg-info/PKG-INFO
writing dependency_links to hare_mp.egg-info/dependency_links.txt
writing entry points to hare_mp.egg-info/entry_points.txt
writing requirements to hare_mp.egg-info/requires.txt
writing top-level names to hare_mp.egg-info/top_level.txt
reading manifest file 'hare_mp.egg-info/SOURCES.txt'
writing manifest file 'hare_mp.egg-info/SOURCES.txt'
running build_ext
main (unittest.loader._FailedTest) ... ERROR
test_both_dix_and_sns_pools_can_exist (test.test_cdf.TestCDF) ... ok
test_disk_refs_can_be_empty (test.test_cdf.TestCDF) ... ok
test_dix_pool_uses_metadata_devices (test.test_cdf.TestCDF) ... ok
test_iface_type_can_be_null (test.test_cdf.TestCDF) ... ok
test_invalid_storage_set_configuration_rejected (test.test_cdf.TestCDF)
This test case checks whether exception will be raise if total ... ok
test_it_works (test.test_cdf.TestCDF) ... ok
test_md_pool_ignored (test.test_cdf.TestCDF) ... ok
test_metadata_is_hardcoded (test.test_cdf.TestCDF) ... ok
test_multiple_nodes_supported (test.test_cdf.TestCDF) ... ok
test_provided_values_respected (test.test_cdf.TestCDF) ... ok
test_template_sane (test.test_cdf.TestCDF) ... ok
test_disks_empty (test.test_cdf.TestTypes) ... ok
test_m0clients (test.test_cdf.TestTypes) ... ok
test_m0server_with_disks (test.test_cdf.TestTypes) ... ok
test_maybe_none (test.test_cdf.TestTypes) ... ok
test_pooldesc_empty (test.test_cdf.TestTypes) ... ok
test_protocol (test.test_cdf.TestTypes) ... ok
test_invalid_machine_id (test.test_validator.TestValidator) ... ok
test_is_cluster_first_node (test.test_validator.TestValidator) ... ok

======================================================================
ERROR: main (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: main
Traceback (most recent call last):
  File "/usr/lib64/python3.6/unittest/loader.py", line 153, in loadTestsFromName
    module = __import__(module_name)
  File "/w/hare/provisioning/miniprov/hare_mp/main.py", line 34, in <module>
    from cortx.utils.product_features import unsupported_features
  File "/usr/lib/python3.6/site-packages/cortx/utils/product_features/unsupported_features.py", line 19, in <module>
    from cortx.utils.data.access import Query
  File "/usr/lib/python3.6/site-packages/cortx/utils/data/access/__init__.py", line 18, in <module>
    from .base_model import BaseModel
  File "/usr/lib/python3.6/site-packages/cortx/utils/data/access/base_model.py", line 16, in <module>
    from schematics.models import Model
ModuleNotFoundError: No module named 'schematics'


----------------------------------------------------------------------
Ran 20 tests in 0.129s

FAILED (errors=1)
Test failed: <unittest.runner.TextTestResult run=20 errors=1 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=20 errors=1 failures=0>
```

knekrasov commented on provisioning/miniprov/hare_mp/cdf.py:
```diff
@@ -187,16 +187,32 @@ def _create_profile_descriptions(
+            self, nodes: List[NodeDesc]) -> List[FdmiFilterDesc]:
+        fdmi_filters: List[FdmiFilterDesc] = []
+
+        fdmi_filters.append(
```
These values don't look like a harmless stub. Per my understanding, this will register non empty FDMI filter that will be applied unconditionally. Given it is CDF generator, this filter will be added every time Mini provisioner deploys Hare.

Is it correct?

max-seagate replied:
Yes, it is correct. I tried to add an empty filters section, but seems like `cdf.py` doesn't support this and I can't avoid adding it because of how dhall works. Would you please help me with this?

knekrasov replied:
@max-seagate I've just pushed a fixup commit to your PR. Please have a look and squash if it looks good.

max-seagate replied:
Thank you. Let me check.

max-seagate replied:
Single node bootstrap succeeds with your patch, so seems like it's good. Thanks.

knekrasov reviewed (commented):
`make check` in `cfgen/` folder fails. Seems like some [test data](https://github.com/Seagate/cortx-hare/tree/main/cfgen/tests) of cfgen should also be updated.

max-seagate commented:
> make check in cfgen/ folder fails. Seems like some test data of cfgen should also be updated.
@knekrasov: now it doesn't. I've added an empty list of FDMI filter groups. Is this enough or you would like to have something more sophisticated?

knekrasov commented:
@max-seagate If `make check` doesn't fail, then we're OK. I don't think we need something more sophisticated. Thanks!

max-seagate commented:
@knekrasov, @mssawant: seems like all review comments are addressed. Is there something else I need to take care of?

knekrasov commented:
Note: this PR blocks https://github.com/Seagate/cortx-motr/pull/736

cc @mehjoshi

knekrasov commented:
@max-seagate I can see that VM deployment has failed by some reason

max-seagate commented:
@knekrasov: the failure http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Hare/395/artifact/deployment_status.log

```
Traceback (most recent call last):
  File \"/opt/seagate/cortx/hare/bin/cfgen\", line 1999, in <module>
    main()
  File \"/opt/seagate/cortx/hare/bin/cfgen\", line 194, in main
    validate_cluster_desc(cluster_desc)
  File \"/opt/seagate/cortx/hare/bin/cfgen\", line 313, in validate_cluster_desc
    validate_fdmi_filters_desc(desc['fdmi_filters'], desc['nodes'])
  File \"/opt/seagate/cortx/hare/bin/cfgen\", line 479, in validate_fdmi_filters_desc
    assert all_unique(x['name'] for x in fdmi_filters_desc), \
TypeError: 'NoneType' object is not iterable
)"
```
The code:
```
 307 def validate_cluster_desc(desc: Dict[str, Any]) -> None:                         
 308     validate_nodes_desc(desc['nodes'])                                           
 309     validate_pools_desc(desc['pools'])                                           
 310     if 'profiles' in desc:                                                       
 311         validate_profiles_desc(desc['profiles'], desc['pools'])                  
 312     if 'fdmi_filters' in desc:                                                   
 313         validate_fdmi_filters_desc(desc['fdmi_filters'], desc['nodes'])          
```

`fdmi_filters` section should not be present in the CDF, but for some reason it is. The tests I know about don't fail on my VM, so obviously something else is going on here. Do you know how to debug such issues?

The VM deployment will fail anyway if Motr is not from at least https://github.com/Seagate/cortx-motr/pull/736.

knekrasov commented:
@max-seagate according to this line

```
"comment": "Command \"/opt/seagate/cortx/hare/bin/hare_setup config --config json:///opt/seagate/cortx_configs/provisioner_cluster.json --file '/var/lib/hare/cluster.yaml'\" run",
```
the CDF is generated by our CDF generator (the `cdf.py` file that you're familiar with already). Which means that either the ConfStore file is not quite valid in respect to this new FDMI filter thing or there is a bug in the CDF generator.

So IMO the central point for debugging is the ConfStore file (we don't know how does it look like in VM Deployment job). Once you get it, you'll be something to debug against.

@vaibhavparatwar do you know (by any chance) if it is possible to get the ConfStore file which is used in VM deployment job? Or maybe you know whom should we ping?

max-seagate commented:
@knekrasov: here is `provisioner_cluster.json` that had been used in the test (I hope this is the one): http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Hare/395/artifact/provisioner_cluster.json

vaibhavparatwar commented:
> @knekrasov: here is `provisioner_cluster.json` that had been used in the test (I hope this is the one): http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Hare/395/artifact/provisioner_cluster.json

Yes @max-seagate this is the one.

max-seagate commented:
@knekrasov: If I run the same command on my local VM I get
```
> /opt/seagate/cortx/hare/bin/hare_setup config --config json:///tmp/provisioner_cluster.json --file '/var/lib/hare/cluster.yaml'
Traceback (most recent call last):
  File "/opt/seagate/cortx/hare/bin/hare_setup", line 33, in <module>
    sys.exit(load_entry_point('hare-mp==2.0.0', 'console_scripts', 'hare_setup')())
  File "/opt/seagate/cortx/hare/bin/hare_setup", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/w/hare/.py3venv/lib64/python3.6/site-packages/importlib_metadata/__init__.py", line 178, in load
    module = import_module(match.group('module'))
  File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/w/hare/.py3venv/lib64/python3.6/site-packages/hare_mp-2.0.0-py3.6.egg/hare_mp/main.py", line 39, in <module>
ModuleNotFoundError: No module named 'hax'
```
Do you know how to run it properly? I'm using `sudo git clean -dfx && make && sudo make install` to build & install Hare.

max-seagate commented:
Even `source ./.py3venv/bin/activate` from a comment above doesn't help.

max-seagate commented:
Debugged and fixed with help from @knekrasov. The trick with `hax` was:
```
cd hax/
python ./setup.py install
```

max-seagate commented:
@mssawant: VM deployment is going to fail anyway because https://github.com/Seagate/cortx-motr/pull/736 is required for this patch to run Motr/Hare cluster. Is there something I need to do because of this?
CC @knekrasov.

max-seagate commented:
Rebased on the latest main & fixed new post-rebase bootstrap issues.

Please use https://github.com/Seagate/cortx-motr/pull/719 if you want to test the patch with Motr.

johnbent commented:
@kaustubh-d and @mbcortx , do we need this merged for ADG work?

huanghua78 commented:
Can the gatekeeper of hare merge this PR? @knekrasov @mssawant 
This is needed for ADG work in this sprint and this PI.

knekrasov commented:
@max-seagate can you please rebase the branch against the latest main and resolve the conflicts?
I can also see that VM deployment is failed (but probably that can be caused by some infra issues we had in the past). I believe rebasing will force the Jenkins job to re-run. Thanks!

huanghua78 commented:
@knekrasov I have rebased the PR on the latest main branch.
Please check.

knekrasov commented:
@huanghua78 VM Deployment 7.9 [has failed](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/CentOS-7.9.2009/job/pr-deployment/267/console). I guess that this error may relate to the changes in this PR:
```
2:02:20      2021-09-14 09:01:55 nodecli ERROR cortx_setup command Failed: salt command failed, reason {
12:02:20              srvnode-1: {
12:02:20          "cmd_|-Stage - Configure Hare_|-__slot__:salt:setup_conf.conf_cmd('/opt/seagate/cortx/hare/conf/setup.yaml', 'hare:config')_|-run": {
12:02:20              "comment": "Command \"/opt/seagate/cortx/hare/bin/hare_setup config --config json:///opt/seagate/cortx_configs/provisioner_cluster.json --file '/var/lib/hare/cluster.yaml'\" run",
12:02:20              "changes": {
12:02:20                  "pid": 54826,
12:02:20                  "retcode": 255,
12:02:20                  "stdout": "",
12:02:20                  "stderr": "2021-09-14 09:01:55,628 [ERROR] Error performing configuration (Command ['cfgen', '-o', '/var/lib/hare', '/var/lib/hare/cluster.yaml'] exited with error code 1. Command output:
12:02:20      [1;33mWarning[0m: Could not locate a cache base directory from the environment.
12:02:20    
12:02:20      You can provide a cache base directory by pointing the $XDG_CACHE_HOME
12:02:20      environment variable to a directory with read and write permissions.
12:02:20    
12:02:20      Traceback (most recent call last):
12:02:20        File \"/opt/seagate/cortx/hare/bin/cfgen\", line 2313, in <module>
12:02:20          main()
12:02:20        File \"/opt/seagate/cortx/hare/bin/cfgen\", line 241, in main
12:02:20          cluster = build_cluster(cluster_desc)
12:02:20        File \"/opt/seagate/cortx/hare/bin/cfgen\", line 2298, in build_cluster
12:02:20          for fdmi_filter in cluster_desc['fdmi_filters']:
12:02:20      TypeError: 'NoneType' object is not iterable
12:02:20      )"
12:02:20              }
12:02:20          }
12:02:20      },
12:02:20       }, args {
12:02:20          "targets": "*",
12:02:20          "fun": "state.apply",
12:02:20          "fun_args": [
12:02:20              "components.hare.config.config"
12:02:20          ],
12:02:20          "fun_kwargs": {},
12:02:20          "kw": {
12:02:20              "tgt_type": "compound",
12:02:20              "full_return": true
12:02:20          },
12:02:20          "secure": false,
12:02:20          "nowait": false
12:02:20      }
12:02:20       -----
12:02:20    stdout_lines: <omitted>
12:02:20  
12:02:20  TASK [factory_vm_deployment : [ Validate Deployment ] : Component Deployment Failure] ***
12:02:20  [WARNING]: conditional statements should not include jinja2 templating
12:02:20  delimiters such as {{ }} or {% %}. Found: {{ FAILED }} == True
12:02:20  fatal: [srvnode-1]: FAILED! => changed=false 
12:02:20    msg: Component deployment failed !!
```

huanghua78 commented:
> @mssawant: VM deployment is going to fail anyway because [Seagate/cortx-motr#736](https://github.com/Seagate/cortx-motr/pull/736) is required for this patch to run Motr/Hare cluster. Is there something I need to do because of this?
> CC @knekrasov.

@max-seagate , do we need to merge  [Seagate/cortx-motr#736](https://github.com/Seagate/cortx-motr/pull/736) and this PR at the same time to avoid such VM deployment failure?
"at the same time," I mean, we should merge them in sync.

huanghua78 commented on cfgen/cfgen:
```diff
@@ -2149,6 +2294,10 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
         ConfProfile.build(conf, root_id, prof)
 
+    if 'fdmi_filters' in cluster_desc:
+        for fdmi_filter in cluster_desc['fdmi_filters']:
```
Here, the "cluster_desc['fdmi_filters']" might be None, because some cdf file may not have this defined.

Can we check if it is None?
```
    if 'fdmi_filters' in cluster_desc and cluster_desc['fdmi_filters'] is not None:
        for fdmi_filter in cluster_desc['fdmi_filters']:
```
Just like the same checking in other places?

I am not sure about this. So I am asking @max-seagate for help.

huanghua78 replied:
I see similar code in:
```
 348 def validate_cluster_desc(desc: Dict[str, Any]) -> None:
 349     validate_nodes_desc(desc['nodes'])
 350     validate_pools_desc(desc['pools'])
 351     if 'profiles' in desc:
 352         validate_profiles_desc(desc['profiles'], desc['pools'])
 353     if 'fdmi_filters' in desc and desc['fdmi_filters'] is not None:
 354         validate_fdmi_filters_desc(desc['fdmi_filters'], desc['nodes'])
 355

```

huanghua78 commented:
I have verified that:
CORTX-Hare: with the branch from this PR.
CORTX-Motr: with the fdmi_int branch from https://github.com/Seagate/cortx-motr/pull/719

With the following CDF file chanegs (enanling fdmi):
```
[520428@ssc-vm-g3-rhev4-0489 ~]$ diff -u singlenode.yaml CDF.yaml
--- singlenode.yaml     2021-09-15 18:12:38.238716194 +0800
+++ CDF.yaml    2021-09-15 18:04:24.681528044 +0800
@@ -47,8 +47,8 @@
 #profiles:
 #  - name: default
 #    pools: [ the pool ]
-#fdmi_filters:
-#  - name: test
-#    node: localhost
-#    client_index: 1
-#    substrings: ["Bucket-Name", "Object-Name", "Object-URI"]
+fdmi_filters:
+  - name: test
+    node: localhost
+    client_index: 1
+    substrings: ["Bucket-Name", "Object-Name", "Object-URI"]

```

I can successfully:
```
sudo hctl bootstrap --mkfs ~/CDF.yaml
```
and then

```
[520428@ssc-vm-g3-rhev4-0489 ~]$ cp /opt/seasudo hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x2f 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x4f 'default': 'the pool' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   192.168.60.64@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.60.64@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.60.64@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x29  192.168.60.64@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2c  192.168.60.64@tcp:12345:4:2
```

mssawant commented:
Hi @max-seagate, @huanghua78, it would have been good if would have tested a deployment using atleast a custom build comprising of respective motr and hare branches just to make sure that main does not fail. Can we do that please? Or if already deployed, can any of you please post the link to the build?

huanghua78 commented:
> Hi @max-seagate, @huanghua78, it would have been good if would have tested a deployment using atleast a custom build comprising of respective motr and hare branches just to make sure that main does not fail. Can we do that please? Or if already deployed, can any of you please post the link to the build? cc @knekrasov, @truptiatseagate

I did the integrated deployment manually. @rajkumarpatel2602 did it manually.

Who can help to do a "custom build"?

suvratjoshi commented:
> Who can help to do a "custom build"?

Hi @huanghua78 
This link can be used to create a custom build :
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/custom-ci/

Once you login to Jenkins using this link, you will need to fill in the branch name and corresponding repo link for the components that have code changes.

Once build is successful, these links can be used to test shingle node and three node deployments using fresh Centos 7.9 VMs.

Single node:
http://eos-jenkins.mero.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-1Node-CentOS-7.9/

Three node:
http://eos-jenkins.mero.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-3Node-CentOS-7.9/

mssawant commented:
Deployments,
1 node VM: http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-1Node-CentOS-7.9/154 : Passed
3 node VM: http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-3Node-CentOS-7.9/103 : Failed, Hare + Motr stack started fine, ha_proxy and other s3 services failing on node-3,
```
07:59:38  changed: [srvnode-1 -> ssc-vm-g3-rhev4-1070.****] => changed=true 
07:59:38    rc: 0
07:59:38    stderr: |-
07:59:38      Shared connection to ssc-vm-g3-rhev4-1070.**** closed.
07:59:38    stderr_lines: <omitted>
07:59:38    stdout: |-
07:59:38      Failed Service : systemd:haproxy
07:59:38      Failed Service : systemd:s3backgroundconsumer
07:59:38      Failed Service : s3server-count
07:59:38    stdout_lines: <omitted>
```

huanghua78 commented:
> Deployments,
> 1 node VM: http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-1Node-CentOS-7.9/154 : Passed
> 3 node VM: http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-3Node-CentOS-7.9/103 : Failed, > Hare + Motr stack started fine, ha_proxy and other s3 services failing on node-3,

Adding more:
@rajkumarpatel2602 has done some basic I/O testing （using m0cp & m0cat) on the single node deployment.
The results are good and correct.

mssawant commented:
Deployment for 3 node cluster passed : http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-3Node-CentOS-7.9/107/
custom-ci Job used : http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/custom-ci/2102/parameters/
